src/shared/ui/survey/StackedCheckboxItem/StackedCheckboxItem.tsx

```diff
@@ -53,41 +50,30 @@ const StackedCheckboxItem: FC<Props> = ({
     });
   }, [rows, textReplacer]);
 
-  const onValueChange = (option: StackedRowItemValue, row: Item) => {
-    const currentRowId = row.id;
-    const optionId = option.id;
+  const isValueSelected = (value: Item, rowIndex: number) => {
+    if (!values || !values[rowIndex]?.length) {
+      return false;
+    }
+
+    const selectedValue = values[rowIndex].find(item => item.id === value.id);
+
+    return !!selectedValue;
+  };
 
+  const onValueChange = (option: StackedRowItemValue, rowIndex: number) => {
+    let newValues: Values = [];
     if (!values) {
-      const newValues = [
-        {
-          rowId: currentRowId,
-          optionIds: [optionId],
-        },
-      ];
-
-      onChange(newValues);
+      newValues.length = config.rows.length;
+      newValues[rowIndex] = [option];
     } else {
-      const currentRow = values.find(({ rowId }) => rowId === currentRowId);
-      const currentOptionIds = currentRow?.optionIds ?? [];
-      let newOptionIds = [...currentOptionIds];
-
-      if (currentOptionIds.includes(optionId)) {
-        newOptionIds = currentOptionIds.filter(id => id !== optionId);
-      } else {
-        newOptionIds = [...newOptionIds, optionId];
-      }
-
-      const newValues = [
-        ...values.filter(({ rowId }) => rowId !== currentRowId),
-        {
-          rowId: currentRowId,
-          optionIds: newOptionIds,
-        },
-      ];
-      const isEmpty = !newValues.some(({ optionIds }) => optionIds.length > 0);
-
-      onChange(isEmpty ? null : newValues);
+      newValues = [...values];
+
+      newValues[rowIndex] = isValueSelected(option, rowIndex)
+        ? [...newValues[rowIndex].filter(value => value.id !== option.id)]
```

Comment:
No need to create a new array via the spread operator because .filter function always creates a new array by default